pkg/wakaama/contrib/objects/on_off_switch.c

```diff
@@ -0,0 +1,492 @@
+ * @param[in, out] data_array   IDs of resources to read. Array of data structures to place values.
+ * @param[in] object            On/off switch object handle
+ *
+ * @return COAP_205_CONTENT on success
+ * @return COAP_404_NOT_FOUND if the instance was not found
+ * @return COAP_500_INTERNAL_SERVER_ERROR otherwise
+ */
+static uint8_t _read_cb(lwm2m_context_t * context, uint16_t instance_id, int * num_data,
+                        lwm2m_data_t ** data_array, lwm2m_object_t * object);
+
+/**
+ * @brief 'Write' callback for the LwM2M On/off switch object implementation.
+ *
+ * @param[in] context           LWM2M Context
+ * @param[in] instance_id       ID of the instance to write resource to.
+ * @param[in] num_data          Number of elements in @p data_array.
+ * @param[in] data_array        IDs of resources to write and values.
+ * @param[in] object            On/off switch object handle
+ *
+ * @return COAP_204_CHANGED on success
+ * @return COAP_404_NOT_FOUND if the instance was not found
+ * @return COAP_400_BAD_REQUEST if a value is not encoded correctly
+ * @return COAP_500_INTERNAL_SERVER_ERROR otherwise
+ */
+static uint8_t _write_cb(lwm2m_context_t * context, uint16_t instance_id, int num_data,
+                         lwm2m_data_t * data_array, lwm2m_object_t * object, lwm2m_write_type_t write_type);
+
+/**
+ * @brief Gets the current value of a given @p instance.
+ *
+ * @param[in, out] data     Initialized data structure.
+ * @param[in] instance      Pointer to the instance to get the value from.
+ *
+ * @return COAP_205_CONTENT on success
+ * @return COAP_404_NOT_FOUND if the value is not found
+ */
+static uint8_t _get_value(lwm2m_data_t *data, lwm2m_obj_on_off_switch_inst_t *instance);
+
+/**
+ * @brief Mark a resource as changed on the LwM2M engine.
+ *
+ * @param[in] instance_id    ID of the instance to mark the resource as changed.
+ * @param[in] resource_id    ID of the resource to mark as changed.
+ */
+static void _mark_resource_changed(uint16_t instance_id, uint16_t resource_id);
+
+struct lwm2m_on_off_switch_object {
+    lwm2m_object_t wakaama_object;                   /**< Wakaama internal object */
+    mutex_t lock;                                    /**< mutex for the instances access */
+    lwm2m_obj_on_off_switch_inst_t *free_instances;  /**< list of free instances */
+    lwm2m_obj_on_off_switch_inst_t instances[CONFIG_LWM2M_ON_OFF_SWITCH_INSTANCES_MAX]; /**< instances */
+};
+
+struct lwm2m_on_off_switch_object _on_off_switch_object = {
+    .lock = MUTEX_INIT,
+    .wakaama_object = {
+        .objID = LWM2M_ON_OFF_SWITCH_OBJECT_ID,
+        .instanceList = NULL,
+        .readFunc = _read_cb,
+        .writeFunc = _write_cb,
+        .executeFunc = NULL,
+        .createFunc = NULL,
+        .deleteFunc = NULL,
+        .discoverFunc = NULL,
+        .userData = NULL
+    }
+};
+
+static uint8_t _get_value(lwm2m_data_t *data, lwm2m_obj_on_off_switch_inst_t *instance)
```

Comment:
I was confused for a second which way 'value is get from'. Perhaps `data` and `instance` are not descriptive enough. (It might also be my lack of familiarity with lwm2m).

Reply:
I extended the function documentation. Instance refers to object instance, common term in LwM2M.